kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/aggregate/AggregateWithImmutableIdentifierFactory.kt

```diff
@@ -0,0 +1,120 @@
+) : AggregateFactory<A> {
+
+    companion object {
+
+        /**
+         * Reified factory method for aggregate factory using string as aggregate identifier.
+         * @return instance of AggregateWithImmutableIdentifierFactory
+         */
+        inline fun <reified A : Any> usingStringIdentifier() = usingIdentifier<A, String>(String::class) { it }
+
+        /**
+         * Factory method for aggregate factory using string as aggregate identifier.
+         * @return instance of AggregateWithImmutableIdentifierFactory
+         */
+        fun <A : Any> usingStringIdentifier(clazz: KClass<A>) = usingIdentifier(clazz, String::class, DefaultString)
+
+        /**
+         * Reified factory method for aggregate factory using UUID as aggregate identifier.
+         * @return instance of AggregateWithImmutableIdentifierFactory
+         */
+        inline fun <reified A : Any> usingUUIDIdentifier() = usingIdentifier<A, UUID>(UUID::class, DefaultUUID::apply)
+
+        /**
+         * Factory method for aggregate factory using UUID as aggregate identifier.
+         * @return instance of AggregateWithImmutableIdentifierFactory
+         */
+        fun <A : Any> usingUUIDIdentifier(clazz: KClass<A>) = usingIdentifier(clazz, UUID::class, DefaultUUID)
+
+        /**
+         * Reified factory method for aggregate factory using specified identifier type and converter function.
+         * @param idClazz identifier class.
+         * @param idExtractor extractor function for identifier from string.
+         * @return instance of AggregateWithImmutableIdentifierFactory
+         */
+        inline fun <reified A : Any, ID : Any> usingIdentifier(idClazz: KClass<ID>, noinline idExtractor: (String) -> ID) =
+            AggregateWithImmutableIdentifierFactory(clazz = A::class, idClazz = idClazz, idExtractor = object : AggregateIdentifierConverter<ID> {
+                override fun apply(it: String): ID = idExtractor(it)
+            })
+
+        /**
+         * Factory method for aggregate factory using specified identifier type and converter.
+         * @param idClazz identifier class.
+         * @param idExtractor extractor for identifier from string.
+         * @return instance of AggregateWithImmutableIdentifierFactory
+         */
+        fun <A : Any, ID : Any> usingIdentifier(aggregateClazz: KClass<A>, idClazz: KClass<ID>, idExtractor: AggregateIdentifierConverter<ID>) =
+            AggregateWithImmutableIdentifierFactory(clazz = aggregateClazz, idClazz = idClazz, idExtractor = idExtractor)
+    }
+
+
+    @Throws(IllegalArgumentException::class)
+    override fun createAggregateRoot(aggregateIdentifier: String, message: DomainEventMessage<*>?): A {
+
+        val constructor = invokeReporting(
+            "The aggregate [${clazz.java.name}] doesn't provide a constructor for the identifier type [${idClazz.java.name}]."
+        ) { clazz.java.getConstructor(idClazz.java) }
```

Comment:
Can a function reference to a factory method be used here instead of hardcoding Constructor lookup via reflection?

```
val constructor = clazz.java.getConstructor(idClazz.java)
val id: ID  = idExtractor.apply(aggregateIdentifier)
constructor.newInstance(id)
``` 
can still be passed to this class as the factory function, but `AggregateWithImmutableIdentifierFactory` could potentially be used without reflection. 
`(ID) -> A` could be used for the factory function, with `idExtractor.apply` being called prior to invoking it.

Reply:
Hi Stefan,
not sure I understood you correctly.

We need to know the `clazz` of the aggregate inside the `AggregateWithImmutableIdentifierFactory`, so I pass it in. Now we will need a factory method to create an instance of it. In generally, if the aggregate is a data class, we should be able to fallback to the constructor, since it must exist in a data class, but for sure could pass it from the outside.

That would mean, that the user of this factory needs to determine it by him-/herself and pass in to the factory.
Probably a good idea. Do you mean this?

Reply:
I implemented as the way I understood you. Now the factory function can be passed from outside.
I decided not to move it it completely into `spring-auto-configure`, because we want not spring guys to use the extension too. So the default of the factory method is the reflective constructor finding and instance creation is still there.

I'll let the PR in draft, since I want more discussion and more feedback.

For example -> in current `axon-spring` aggregates are prototype-scoped spring beans. In this extension, they are not... I even started to experiment with this, but it seems black magic and I doubt it is really needed. The only reason for doing so I found so far was the scan for aggregates and I do a manual classpath scan to find those.

Is there a use case we would benefit from getting aggregates be prototype spring beans (despite the detection by spring)?